docs/topics/db/optimization.txt

```diff
@@ -261,6 +261,28 @@ But:
 
 .. _overuse_of_count_and_exists:
 
+
+Don't overuse ``contains()``
+------------------------------------------
+
+If you need to check if an object is a member of a queryset then ``contains``
+can cause a new DB call for each usage. This is potentially 
+inefficient and can introduce an n+1 issue.
+
+For example::
+    for user in User.objects.all():
+        if not post.authors.contains(user):
+            print(f'{user.name} is not an author of this blog post')
+
+This implementation causes 1 call for all users, and then another
+for each time contains is called.
```

Comment:
Thanks for taking this on 👍  The example you have given here is a good one insomuch as it's clear and correct; unfortunately - it's just not what this ticket is about though. I'll try and clarify below:

The ticket is to adapt the existing section about `count()` and `exists()` to also include `contains()` - not to introduce a new separate section.

The new warning about `contains()` should also be about (just as it is for `count()` and `exists()` ) not making 2 queries when 1 will do. Specifically, "don't make a query if later on you're going to make a second query that will contain the information you requested in the first query". This is a distinct problem to the (n+1) issue that your example brings up. Both are perfectly valid things to warn against, but the former issue is perhaps an easier trap to fall into, which is why it merits its own section in the docs.

The issue your example discusses, also falls into a broader category of issues, which are not specific to `contains()` - think caching, prefetching etc. If you think that the docs need an example like the one you have added here, feel free to create a different ticket for that :)


If you wanted to keep working on this ticket, you'll probably need to adapt the current example, since it doesn't lend itself very well to `contains`. One such example could be displaying a list of users that are part of a group and that also says something along the lines of:
 - "There are 6 members in this group"
 - "You and 5 other users are members of this group"
 - "There are no members in this group"
 
(depending on who the members are) 

This is an example where someone might think about using all 3 of `count`, `exists` and `contains` but there's no point because later on down the code they're going to need to evaluate the full list anyway. (There may well be a better example to use - but I think this is the kind of thing the ticket is getting at. I hope this helps).